File: python/ionveil/tests/stress/hyper_matrix_test.py
import hashlib
import unittest

from ionveil.models import (
    DispatchOrder, VesselManifest, merge_manifests,
    classify_severity, create_batch_orders, SLA_BY_SEVERITY,
)
from ionveil.dispatch import (
    plan_dispatch, dispatch_batch, has_conflict, BerthSlot,
    estimate_cost, estimate_turnaround, check_capacity,
    mutual_aid_required,
)
from ionveil.routing import (
    choose_route, channel_score, estimate_transit_time, plan_multi_leg, Waypoint,
    build_adjacency_costs, haversine_distance, find_cheapest_route,
)
from ionveil.policy import (
    next_policy, previous_policy, should_deescalate, check_sla_compliance,
    sla_percentage, all_policies, policy_index, escalation_chain,
)
from ionveil.queue import should_shed, queue_health, estimate_wait_time, PriorityQueue, drain_by_priority
from ionveil.resilience import (
    replay_events, deduplicate, replay_converges,
)
from ionveil.statistics import percentile, mean, moving_average, exponential_moving_average, compute_breach_rate
from ionveil.workflow import (
    can_transition, is_terminal_state, is_valid_state, shortest_path,
)
from ionveil.security import (
    verify_signature, sign_manifest, verify_manifest, sanitise_path,
)
from shared.contracts.contracts import get_service_url, validate_contract, topological_order

TOTAL_CASES = 12400


def _make_case(idx: int) -> None:
    # Large-scale bug pressure: half the matrix targets known unresolved invariants.
    # This keeps the baseline sparse for apex-level training while still grounding
    # failures in concrete business rules.
    if idx % 4 == 0:
        assert next_policy("normal", 2) == "watch"
        return

    if idx % 4 == 1:
        replayed = replay_events(
            [
                {"id": "x", "sequence": 1},
                {"id": "x", "sequence": 4},
                {"id": "y", "sequence": 2},
            ]
        )
        assert [f"{e['id']}:{e['sequence']}" for e in replayed] == ["y:2", "x:4"]
        return

    # --- Sub-categories: each exercises a distinct source-code bug ---
    severity_a = (idx % 5) + 1
    severity_b = ((idx * 3) % 5) + 1
    sla_a = 20 + (idx % 90)
    sla_b = 20 + ((idx * 2) % 90)

    a = DispatchOrder(f"a-{idx}", severity_a, sla_a)
    b = DispatchOrder(f"b-{idx}", severity_b, sla_b)

    sub = (idx // 4) % 16

    if sub == 0:
        # Bug: urgency_score uses severity * 8 instead of * 10
        expected = severity_a * 10 + max(0, 120 - sla_a)
        assert a.urgency_score() == expected
        return

    if sub == 1:
        # Bug: choose_route picks highest latency instead of lowest
        routes_list = [
            {"channel": "alpha", "latency": 2 + (idx % 9)},
            {"channel": "beta", "latency": 1 + (idx % 3)},
            {"channel": "gamma", "latency": 4 + (idx % 4)},
        ]
        route = choose_route(routes_list, set())
        expected_latency = min(r["latency"] for r in routes_list)
        assert route["latency"] == expected_latency
        return

    if sub == 2:
        # Bug: replay_events keeps LOWER sequence instead of higher
        seq_high = 10 + (idx % 20)
        replayed = replay_events([
            {"id": f"ev-{idx}", "sequence": 1},
            {"id": f"ev-{idx}", "sequence": seq_high},
            {"id": f"other-{idx}", "sequence": 3},
        ])
        ev_entry = [e for e in replayed if e["id"] == f"ev-{idx}"][0]
        assert ev_entry["sequence"] == seq_high
        return

    if sub == 3:
        # Bug: should_shed uses > instead of >= (off-by-one)
        limit = 30 + (idx % 20)
        assert should_shed(limit, limit, False) is True
        return

    if sub == 4:
        # Bug: estimate_cost uses 0.45 instead of 0.5 per km
        urg = (idx % 5) + 1
        cost_near = estimate_cost(urg, 100.0)
        cost_far = estimate_cost(urg, 200.0)
        per_km = round((cost_far - cost_near) / 100.0, 2)
        assert per_km == 0.5
        return

    if sub == 5:
        # Bug: compute_breach_rate uses integer division (//)
        base = 50.0 + float(idx % 40)
        times = [base, 150.0, 200.0 + float(idx % 30)]
        rate = compute_breach_rate(times, 100.0)
        expected = round(2 / 3, 4)
        assert abs(rate - expected) < 0.001
        return

    if sub == 6:
        # Bug: EMA uses values[i-1] instead of values[i]
        step = 10.0 + float(idx % 20)
        vals = [1.0, step, step]
        ema = exponential_moving_average(vals, 0.5)
        expected_ema1 = 0.5 * vals[1] + 0.5 * vals[0]
        assert abs(ema[1] - expected_ema1) < 0.01
        return

    if sub == 7:
        # Bug: merge_manifests doesn't deduplicate shared orders
        shared = DispatchOrder(f"shared-{idx}", severity_a, sla_a)
        m_a = VesselManifest(f"A-{idx}", [shared, a])
        m_b = VesselManifest(f"B-{idx}", [shared, b])
        merged = merge_manifests(m_a, m_b)
        unique_ids = {o.id for o in merged.orders}
        assert merged.order_count() == len(unique_ids)
        return

    if sub == 8:
        # Bug: queue_health uses >= EMERGENCY_RATIO for "critical" (should be >)
        limit = 5 * (1 + (idx % 20))
        depth = limit * 4 // 5
        health = queue_health(depth, limit)
        assert health.status == "warning"
        return

    if sub == 9:
        # Bug: build_adjacency_costs only fills upper triangle (asymmetric)
        lat = float(idx % 50)
        lon = float(idx % 80)
        wps = [
            Waypoint("P", lat, lon),
            Waypoint("Q", lat + 2.0 + float(idx % 5), lon + 3.0),
        ]
        matrix = build_adjacency_costs(wps)
        assert abs(matrix[0][1] - matrix[1][0]) < 0.01
        return

    if sub == 10:
        # Bug: mutual_aid_required checks < 2 instead of < 3 for critical
        assert mutual_aid_required(5, 2) is True
        return

    if sub == 11:
        # Bug: drain_by_priority uses > instead of >= (off-by-one)
        q = PriorityQueue(capacity=100)
        threshold = 3 + (idx % 5)
        q.enqueue({"id": f"exact-{idx}", "priority": threshold})
        q.enqueue({"id": f"above-{idx}", "priority": threshold + 2})
        q.enqueue({"id": f"below-{idx}", "priority": threshold - 1})
        drained = drain_by_priority(q, threshold)
        drained_ids = {d["id"] for d in drained}
        assert f"exact-{idx}" in drained_ids
        assert f"above-{idx}" in drained_ids
        return

    if sub == 12:
        # Bug: haversine_distance uses cos(lat1)^2 instead of cos(lat1)*cos(lat2)
        import math
        lat1 = float(idx % 60)
        lon1 = float(idx % 90)
        lat2 = lat1 + 10.0 + float(idx % 20)
        lon2 = lon1 + 15.0
        dist = haversine_distance(lat1, lon1, lat2, lon2)
        R = 6371.0
        lat1_r, lat2_r = math.radians(lat1), math.radians(lat2)
        dlat = math.radians(lat2 - lat1)
        dlon = math.radians(lon2 - lon1)
        a_val = math.sin(dlat / 2) ** 2 + math.cos(lat1_r) * math.cos(lat2_r) * math.sin(dlon / 2) ** 2
        c_val = 2 * math.atan2(math.sqrt(a_val), math.sqrt(1 - a_val))
        expected = round(R * c_val, 2)
        assert abs(dist - expected) < 1.0
        return

    if sub == 13:
        # Bug: should_deescalate uses >= instead of > at exact threshold
        policy_level = ["halted", "restricted", "watch"][idx % 3]
        thresholds = {"halted": 20, "restricted": 10, "watch": 5}
        threshold = thresholds[policy_level]
        assert should_deescalate(threshold, policy_level) is False
        return

    if sub == 14:
        # Bug: escalation_chain returns ascending for de-escalation (should be descending)
        pairs = [("halted", "normal"), ("restricted", "normal"), ("halted", "watch")]
        current, target = pairs[idx % 3]
        chain = escalation_chain(current, target)
        assert len(chain) >= 2
        assert chain[0] == current
        assert chain[-1] == target
        return

    if sub == 15:
        # Bug: find_cheapest_route treats missing cost as 0 (cheapest)
        cost = 10 + (idx % 50)
        routes = [
            {"channel": f"known-{idx}", "cost": cost},
            {"channel": f"unknown-{idx}"},
        ]
        cheapest = find_cheapest_route(routes)
        assert cheapest["channel"] == f"known-{idx}"
        return


class HyperMatrixTest(unittest.TestCase):
    pass


def _add_case(i: int) -> None:
    def test_fn(self: unittest.TestCase) -> None:
        _make_case(i)

    setattr(HyperMatrixTest, f"test_hyper_matrix_{i:05d}", test_fn)


for _i in range(TOTAL_CASES):
    _add_case(_i)


if __name__ == "__main__":
    unittest.main()

File: kotlin/nimbusflow/src/test/kotlin/com/terminalbench/nimbusflow/HyperMatrixTest.kt
package com.terminalbench.nimbusflow

import java.security.MessageDigest
import java.util.stream.IntStream
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource
import kotlin.math.abs
import kotlin.math.ceil

class HyperMatrixTest {
    companion object {
        @JvmStatic
        fun cases(): IntStream = IntStream.range(0, 9200)

        private fun digest(payload: String): String {
            val bytes = MessageDigest.getInstance("SHA-256").digest(payload.toByteArray())
            return bytes.joinToString("") { "%02x".format(it) }
        }
    }

    @ParameterizedTest(name = "hyper-matrix-{0}")
    @MethodSource("cases")
    fun hyperMatrixCase(idx: Int) {
        val group = idx % 12
        val severityA = (idx % 5) + 1
        val severityB = ((idx * 3) % 5) + 1
        val slaA = 20 + (idx % 90)
        val slaB = 20 + ((idx * 2) % 90)

        val orderA = DispatchOrder("a-$idx", severityA, slaA)
        val orderB = DispatchOrder("b-$idx", severityB, slaB)

        // ============================================================
        // GROUP 0: urgencyScore formula + planDispatch sort direction
        // ============================================================
        if (group == 0) {
            // Bug: urgencyScore uses '-' instead of '+'
            val expectedScore = (orderA.urgency * 10) + maxOf(0, 120 - orderA.slaMinutes)
            assertEquals(expectedScore, orderA.urgencyScore(),
                "urgencyScore(urgency=${orderA.urgency}, sla=${orderA.slaMinutes})")

            // Bug: planDispatch sorts ascending instead of descending
            val planned = Allocator.planDispatch(
                listOf(
                    DispatchOrder("hi-$idx", 5, 30),
                    DispatchOrder("lo-$idx", 1, 30),
                ),
                2
            )
            assertEquals("hi-$idx", planned[0].id,
                "planDispatch must place highest urgency first")
        }

        // ============================================================
        // GROUP 1: estimateCost severity scaling + channelScore latency
        // ============================================================
        if (group == 1) {
            // Bug: estimateCost divides by severity instead of multiplying
            val lowSevCost = Allocator.estimateCost(1, 60, 100.0)
            val highSevCost = Allocator.estimateCost(5, 60, 100.0)
            assertTrue(highSevCost > lowSevCost,
                "higher severity must cost more: $highSevCost > $lowSevCost")

            // Bug: channelScore multiplies by latency instead of dividing
            val lowLatScore = Routing.channelScore(5, 0.9, 3)
            val highLatScore = Routing.channelScore(50, 0.9, 3)
            assertTrue(lowLatScore > highLatScore,
                "lower latency must score higher: $lowLatScore > $highLatScore")
        }

        // ============================================================
        // GROUP 2: estimateRouteCost + estimateTurnaround
        // ============================================================
        if (group == 2) {
            // Bug: estimateRouteCost subtracts portFee instead of adding
            val routeCost = Routing.estimateRouteCost(100.0, 2.0, 50.0)
            assertEquals(250.0, routeCost,
                "route cost = distance*fuelRate + portFee")

            // Bug: estimateTurnaround uses floor instead of ceil
            val containers = 501 + (idx % 499)
            val expectedHours = maxOf(1.0, ceil(containers / 500.0))
            val actualHours = Allocator.estimateTurnaround(containers, false)
            assertEquals(expectedHours, actualHours,
                "turnaround for $containers containers")
        }

        // ============================================================
        // GROUP 3: checkCapacity boundary + hasConflict overlap
        // ============================================================
        if (group == 3) {
            // Bug: checkCapacity uses < instead of <=
            assertTrue(Allocator.checkCapacity(100, 100),
                "demand == capacity should be allowed")
            assertTrue(Allocator.checkCapacity(50, 100),
                "demand < capacity should be allowed")
            assertFalse(Allocator.checkCapacity(101, 100),
                "demand > capacity should be rejected")

            // Bug: hasConflict inverts overlap condition
            val slotA = BerthSlot("B1", 8, 14, true, "V1")
            val slotB = BerthSlot("B1", 10, 16, true, "V2")
            assertTrue(BerthPlanner.hasConflict(slotA, slotB),
                "overlapping berth slots must conflict")
            // Non-overlapping: end of first <= start of second
            val slotC = BerthSlot("B1", 8, 10, true, "V3")
            val slotD = BerthSlot("B1", 10, 14, true, "V4")
            assertFalse(BerthPlanner.hasConflict(slotC, slotD),
                "adjacent non-overlapping slots must not conflict")
        }

        // ============================================================
        // GROUP 4: nextPolicy escalation + shouldDeescalate threshold
        // ============================================================
        if (group == 4) {
            // Bug: nextPolicy condition inverted (> instead of <=)
            assertEquals("watch", Policy.nextPolicy("normal", 3),
                "high failure burst should escalate normal→watch")
            assertEquals("restricted", Policy.nextPolicy("watch", 5),
                "high failure burst should escalate watch→restricted")

            // Bug: shouldDeescalate uses > instead of >=
            assertTrue(Policy.shouldDeescalate(10, "halted"),
                "10 successes should deescalate halted (threshold=10)")
            assertFalse(Policy.shouldDeescalate(9, "halted"),
                "9 successes should not deescalate halted")
            assertTrue(Policy.shouldDeescalate(5, "restricted"),
                "5 successes should deescalate restricted (threshold=5)")
            assertTrue(Policy.shouldDeescalate(3, "watch"),
                "3 successes should deescalate watch (threshold=3)")
        }

        // ============================================================
        // GROUP 5: deescalate history + emergency shed ratio
        // ============================================================
        if (group == 5) {
            // Bug: PolicyEngine.deescalate adds history even when no change
            val eng = PolicyEngine()
            eng.deescalate()
            eng.deescalate()
            assertEquals(0, eng.getHistory().size,
                "deescalating 'normal' should not add history entries")

            // Bug: shouldShed uses 0.5 instead of EMERGENCY_RATIO (0.8)
            val hardLimit = 100
            val threshold = (hardLimit * QueueGuard.EMERGENCY_RATIO).toInt()
            assertFalse(QueueGuard.shouldShed(threshold - 1, hardLimit, true),
                "below emergency threshold should not shed")
            assertTrue(QueueGuard.shouldShed(threshold, hardLimit, true),
                "at emergency threshold should shed")
        }

        // ============================================================
        // GROUP 6: estimateWaitTime + variance formula
        // ============================================================
        if (group == 6) {
            // Bug: estimateWaitTime multiplies instead of dividing
            val depth = 10 + (idx % 20)
            val rate = 2.0 + (idx % 3)
            val expected = depth.toDouble() / rate
            val actual = QueueGuard.estimateWaitTime(depth, rate)
            assertTrue(abs(actual - expected) < 0.001,
                "wait time = depth/rate: expected $expected, got $actual")

            // Bug: variance uses n-1 instead of n (population variance)
            val data = listOf(2.0, 4.0, 4.0, 4.0, 5.0, 5.0, 7.0, 9.0)
            val v = Statistics.variance(data)
            assertTrue(abs(v - 4.0) < 0.01,
                "population variance should be 4.0, got $v")
        }

        // ============================================================
        // GROUP 7: token expiry boundary + replay keeps latest
        // ============================================================
        if (group == 7) {
            // Bug: TokenStore.validate uses <= instead of < for expiry
            val store = TokenStore()
            store.store("u-$idx", "tok-$idx", 1000, 60)
            assertFalse(store.validate("u-$idx", "tok-$idx", 1060),
                "token at exact expiry (issuedAt+ttl) should be invalid")
            assertTrue(store.validate("u-$idx", "tok-$idx", 1059),
                "token before expiry should be valid")

            // Bug: replay keeps oldest sequence instead of latest
            val replayed = Resilience.replay(listOf(
                ReplayEvent("r-$idx", 1),
                ReplayEvent("r-$idx", 5),
            ))
            assertEquals(1, replayed.size, "replay should deduplicate by id")
            assertEquals(5, replayed[0].sequence,
                "replay must keep latest sequence, not oldest")
        }

        // ============================================================
        // GROUP 8: percentile formula + severity classify
        // ============================================================
        if (group == 8) {
            // Bug: percentile rank formula produces wrong index
            val p50 = Statistics.percentile(listOf(1, 4, 7, 9), 50)
            assertEquals(4, p50,
                "50th percentile of [1,4,7,9] should be 4")

            // Bug: Severity.classify missing "emergency"→CRITICAL, "urgent"→HIGH
            assertEquals(Severity.CRITICAL, Severity.classify("emergency alert"),
                "emergency should classify as CRITICAL")
            assertEquals(Severity.HIGH, Severity.classify("urgent repair"),
                "urgent should classify as HIGH")
        }

        // ============================================================
        // GROUP 9: heatmap negative coords + workflow graph
        // ============================================================
        if (group == 9) {
            // Bug: negative coordinates mapped to (0,0) instead of being skipped
            val heatEvents = listOf(
                HeatmapEvent(-1.0, -1.0, 5.0),
                HeatmapEvent(0.0, 0.0, 1.0)
            )
            val cells = HeatmapGenerator.generate(heatEvents, 3, 3)
            val origin = cells.first { it.row == 0 && it.col == 0 }
            assertEquals(1.0, origin.value,
                "negative coords must be excluded from heatmap")

            // Bug: workflow graph allows departed→allocated (backward transition)
            assertFalse(Workflow.canTransition("departed", "allocated"),
                "departed should not transition backward to allocated")
        }

        // ============================================================
        // GROUP 10: terminal states + service URL format
        // ============================================================
        if (group == 10) {
            // Bug: "cancelled" not in terminalStates
            assertTrue(Workflow.isTerminalState("cancelled"),
                "cancelled must be a terminal state")
            assertTrue(Workflow.isTerminalState("arrived"),
                "arrived must be a terminal state")
            assertFalse(Workflow.isTerminalState("queued"),
                "queued must not be a terminal state")

            // Bug: getServiceUrl missing port in URL
            val url = ServiceRegistry.getServiceUrl("gateway")
            assertNotNull(url)
            assertTrue(url!!.contains(":8160"),
                "service URL must include port: $url")
            assertTrue(url.contains("/health"),
                "service URL must include health path: $url")
        }

        // ============================================================
        // GROUP 11: PQ concurrency + cross-module integration
        // ============================================================
        if (group == 11) {
            // Bug: PriorityQueue size check outside lock
            val pq = PriorityQueue(5)
            for (i in 0 until 5) pq.enqueue(QueueItem("q$i", i))
            assertFalse(pq.enqueue(QueueItem("overflow", 99)),
                "enqueue must reject when at capacity")
            assertEquals(5, pq.size,
                "queue size must not exceed hard limit")

            // Cross-module integration: dispatch + route + workflow
            val orders = Allocator.planDispatch(
                listOf(
                    DispatchOrder("integ-hi-$idx", 5, 15),
                    DispatchOrder("integ-lo-$idx", 1, 120)
                ),
                1
            )
            assertEquals(1, orders.size)
            assertEquals(5, orders[0].urgency,
                "integration: dispatch must select highest urgency")

            val route = Routing.chooseRoute(
                listOf(Route("fast", 2), Route("slow", 20)),
                emptySet()
            )
            assertNotNull(route)
            assertEquals("fast", route!!.channel,
                "integration: routing must select lowest latency")

            assertTrue(Workflow.canTransition("queued", "allocated"))
            assertTrue(Workflow.canTransition("allocated", "departed"))
            assertTrue(Workflow.canTransition("departed", "arrived"))
        }

        // ============================================================
        // DEPENDENCY CHAINS — cross-module integration (even idx only)
        // Each chain tests 2 bugs from DIFFERENT modules than the group.
        // A test passes only when its group bugs AND chain bugs are fixed.
        // Odd idx tests have no chain, giving partial progress per bug pair.
        // ============================================================
        if (idx % 2 == 0) {
            when ((idx / 24) % 6) {
                0 -> {
                    // Needs: estimateCost (*severity) + channelScore (/latency)
                    assertTrue(Allocator.estimateCost(5, 30, 10.0) > Allocator.estimateCost(1, 30, 10.0),
                        "dep-chain-0: higher severity must cost more")
                    assertTrue(Routing.channelScore(5, 0.9, 3) > Routing.channelScore(50, 0.9, 3),
                        "dep-chain-0: lower latency must score higher")
                }
                1 -> {
                    // Needs: nextPolicy (<=1 guard) + deescalate (no-op guard)
                    val pe = PolicyEngine()
                    pe.escalate(5)
                    assertEquals("watch", pe.currentPolicy, "dep-chain-1: escalate past normal")
                    pe.deescalate(); pe.deescalate()
                    assertTrue(pe.getHistory().size <= 2, "dep-chain-1: no-op deescalate silent")
                }
                2 -> {
                    // Needs: estimateWaitTime (/) + variance (population)
                    val w = (1..4).map { QueueGuard.estimateWaitTime(it * 10, 2.0) }
                    assertTrue(abs(Statistics.variance(w) - 31.25) < 0.01,
                        "dep-chain-2: wait+variance = 31.25, got ${Statistics.variance(w)}")
                }
                3 -> {
                    // Needs: estimateRouteCost (+portFee) + estimateTurnaround (ceil)
                    assertEquals(250.0, Routing.estimateRouteCost(100.0, 2.0, 50.0),
                        "dep-chain-3: route cost adds port fee")
                    assertEquals(2.0, Allocator.estimateTurnaround(750, false),
                        "dep-chain-3: turnaround uses ceil")
                }
                4 -> {
                    // Needs: replay (keep latest) + percentile (rank formula)
                    val r = Resilience.replay(listOf(ReplayEvent("dc-$idx", 1), ReplayEvent("dc-$idx", 10)))
                    assertEquals(10, r[0].sequence, "dep-chain-4: replay keeps latest")
                    assertEquals(4, Statistics.percentile(listOf(4, 1, 9, 7), 50),
                        "dep-chain-4: percentile rank")
                }
                5 -> {
                    // Needs: shouldShed (EMERGENCY_RATIO) + sanitisePath (backslash)
                    val hl = 100; val thresh = (hl * QueueGuard.EMERGENCY_RATIO).toInt()
                    assertFalse(QueueGuard.shouldShed(thresh - 1, hl, true),
                        "dep-chain-5: below emergency threshold")
                    assertTrue(QueueGuard.shouldShed(thresh, hl, true),
                        "dep-chain-5: at emergency threshold")
                    assertFalse(Security.sanitisePath("..\\safe.txt").contains(".."),
                        "dep-chain-5: backslash traversal sanitised")
                }
            }
        }

        // ============================================================
        // Shared invariants (should always pass regardless of bugs)
        // ============================================================

        // DispatchBatch partition invariant
        val (bPlanned, bRejected) = Allocator.dispatchBatch(listOf(orderA, orderB), 1)
        assertEquals(2, bPlanned.size + bRejected.size,
            "dispatch batch must partition all orders")

        // Order validation
        if (severityA in 1..5) {
            assertNull(OrderFactory.validateOrder(orderA),
                "order with valid severity should pass validation")
        }

        // Route selection
        val blocked = if (idx % 5 == 0) setOf("beta") else emptySet()
        val route = Routing.chooseRoute(
            listOf(
                Route("alpha", 2 + (idx % 9)),
                Route("beta", idx % 3),
                Route("gamma", 4 + (idx % 4))
            ),
            blocked
        )
        assertNotNull(route, "chooseRoute must find at least one non-blocked route")
        if ("beta" in blocked) {
            assertTrue(route!!.channel != "beta",
                "blocked channels must be excluded from route selection")
        }

        // Workflow state machine invariants
        assertTrue(Workflow.canTransition("queued", "allocated"))
        assertFalse(Workflow.canTransition("arrived", "queued"),
            "terminal state cannot transition")
        assertTrue(Workflow.isTerminalState("arrived"))
        assertFalse(Workflow.isTerminalState("queued"))

        // Deduplicate invariant
        val deduped = Resilience.deduplicate(
            listOf(
                ReplayEvent("d-$idx", 1),
                ReplayEvent("d-$idx", 2),
                ReplayEvent("e-$idx", 1)
            )
        )
        assertEquals(2, deduped.size,
            "deduplicate must keep one per id")

        // Security digest determinism
        if (idx % 17 == 0) {
            val payload = "manifest:$idx"
            val sig = digest(payload)
            assertTrue(Security.verifySignature(payload, sig, sig),
                "signature of identical digests must verify")
            assertFalse(Security.verifySignature(payload, sig.drop(1), sig),
                "truncated signature must fail verification")

            val manifestSig = Security.signManifest(payload, "key")
            assertTrue(Security.verifyManifest(payload, "key", manifestSig),
                "manifest signature round-trip must verify")
        }
    }
}
